File: Deepspace/lib/PowerEfficientSpaceStation.rb
#encoding:utf-8

require_relative 'SpaceStation'
require_relative 'Transformation'
#require_relative 'SuppliesPackage'


module Deepspace

# Power efficient space station: gives more shot power and protection than a usual space station
#
# @author Miguel Ángel Fernández Gutiérrez, Sergio Quijano Rey
class PowerEfficientSpaceStation < SpaceStation

    # @!attribute [Float] efficiency factor for power efficient space station
    @@EFFICIENCYFACTOR = 1.10

    # Class initializer
    # @param station [SpaceStation] basic space station to convert to efficient
    # --Overriden
    def initialize(station)
        # WIP - otra forma de hacerlo sin usar SuppliesPackage ?
        #       Creía bastante importante el usar el super aquí...
        super(station.name, SuppliesPackage.new(station.ammoPower, station.fuelUnits, station.shieldPower))
        @nMedals = station.nMedals
        @pendingDamage = station.pendingDamage
        @weapons = station.weapons
        @shieldBoosters = station.shieldBoosters
        @hangar = station.hangar
    end
    
    # Makes a shot
    # @return [Float] the shot power
    # --Overriden
    def fire
        return super*@@EFFICIENCYFACTOR
    end

    # Use protection shield
    # @return [Float] the shield's energy
    # --Overriden
    def protection
        return super*@@EFFICIENCYFACTOR
    end

    # Receives a loot
    # @param [Loot] loot to be received
    # @return [Transformation] if loot received was efficient (GETEFFICIENT)
    #         or not (NOTRANSFORM)
    # --Overriden
    def setLoot(loot)
        super
        
        if loot.efficient
            return Transformation::GETEFFICIENT
        else
            return Transformation::NOTRANSFORM
        end
    end

end # class PowerEfficientSpaceStation

end # module Deepspace

File: deepspace/src/deepspace/PowerEfficientSpaceStation.java
package deepspace;

/**
 * Class to represent a PowerEfficientSpaceStation
 * 
 * A PowerEfficientSpaceStation is the same as a standar SpaceStation, but its
 * protections and fires are incremented by a fixed factor
 *
 * A PowerEfficientSpaceStation cannot transform to SpaceCity, but can convert to BetaPowerEfficientSpaceStation
 *
 *
 * @author Sergio Quijano Rey
 * */
class PowerEfficientSpaceStation extends SpaceStation{
    // Private attributes
    //==========================================================================
    private final float EFFICIENCYFACTOR = (float) 1.10;
    
    // Constructors
    //==========================================================================
    public PowerEfficientSpaceStation(SpaceStation station){
        super(station);
    }
    
    // Getters
    //==========================================================================
    /**
     * Gets the string representation
     * @return the string representation
     * */
    @Override
    public String toString(){
        return  "EFFICIENTSTATION\n[[" + super.toString() + 
                "EFFICIENCYFACTOR = " + EFFICIENCYFACTOR +
                "\n]]";
    }

    // Setters
    //==========================================================================
    /**
     * The station fires with its power incremented by EFFICIENCYFACTOR factor
     * @return the power of the shot
     * */
    @Override
    public float fire(){
        return super.fire() * EFFICIENCYFACTOR;

    }

    /**
     * The station protects with its shield power incremented by EFFICIENCYFACTOR factor
     * @return the power of the protection
     * */
    @Override
    public float protection(){
        return super.protection() * EFFICIENCYFACTOR;
    }

    /**
     * The station receives a loot
     * We avoid transformations
     * @param lo, the received loot
     * @return Transformation.GETEFFICIENT, if we are going to transform to a BetaPowerEfficientSpaceStation
     *         Transformation.NWEAPONSPROB, otherwise
     * */
    @Override
    public Transformation setLoot(Loot lo){
        Transformation trans = super.setLoot(lo);

        // We check if we transform to BetaPowerEfficientSpaceStation
        if(trans == Transformation.GETEFFICIENT){
            return Transformation.GETEFFICIENT;
        }else{
            return Transformation.NOTRANSFORM;
        }
    }


}
